### ISSUE TITLE
`SourceDirectorySet.srcDir(configurableFileTree)` doesn't support `ConfigurableFileTree` as argument

### ISSUE THREAD

**Issue by antohaby** [gradle/gradle#27882]
### Current Behavior

Passing ConfigurableFileTree of some Directory to `SourceDirectorySet.srcDir` makes the latter iterate through all files of the passed FileTree. And thus failing with following error:

`> Source directory '<first file from the passed FileTree>' is not a directory.`

The current workaround is to create a lazy FileCollection that returns `fileTree.dir` and keeps its task dependencies.
like this: 

```kotlin
val srcDirToAdd = project.files({ fileTree.dir }).buildBy(fileTree)
sourceDirectorySet.srcDir(srcDirToAdd)
```

It is important that getting `fileTree.dir` should be wrapped in lazy provider because accessing dir cant unexpectedly 
materialize underlying provider.

### Expected Behavior

From a [documentation](https://docs.gradle.org/8.5/userguide/working_with_files.html#sec:file_trees) point of view a FileTree can represent a directory of some files. In this context, I'd expect that passing a FileTree as an argument to `SourceDirectorySet.srcDir` would be treated as a Directory + all filtering logic applied in the passed FilleTree.

### Steps to Reproduce

Here is the sample `gradle.build.kts` that reproduces the problem:

```

abstract class MyTask : DefaultTask() {
    @get:Inject
    abstract val objects: ObjectFactory

    @Suppress("LeakingThis")
    @get:InputFiles
    val sources: SourceDirectorySet = objects.sourceDirectorySet("mysources", "MySources")

    @TaskAction
    fun action() {
        println("Sources:")
        println(sources.files.joinToString("\n"))
    }
}

abstract class GeneratorTask : DefaultTask() {
    @get:OutputDirectory
    abstract val outputDirectory: DirectoryProperty

    @TaskAction
    fun action() {
        outputDirectory.get().asFile.mkdirs()
        outputDirectory.file("MyFile.kt").get().asFile.writeText("""fun myFunction() {}""")
    }
}

val generatorTask = tasks.register<GeneratorTask>("generator") {
    outputDirectory.set(layout.buildDirectory.dir("generated"))
}

// Some custom build logic that maps generatorTask output into ConfigurableFileTree
val fileTree = objects.fileTree()
    .from(generatorTask.map { it.outputDirectory.get() })
    .builtBy(generatorTask) // is required because of #27881

val wrappedFileTree = objects.fileCollection()
    .from({ fileTree.dir })
    .builtBy(fileTree)

val myTask = tasks.register<MyTask>("myTask") {
    sources.srcDir(fileTree) // doesn't work
//    sources.srcDirs(wrappedFileTree) // works
}
```

### Gradle version

8.5

**Comment by big-guy** [comment]
I agree that this is weird. The problem is that a FileTree is a FileCollection, so `srcDir(FileCollection)` seems to mean "add all of these directories". My expectation is that this should have meant "add one directory" and failed when there are multiple directories provided. I think this is unintentional. 

I think I would avoid using `SourceDirectorySet` directly on a task. There are some other bits of its API that don't quite make sense (`compiledBy`, `classesDirectory` and `destinationDirectory`) on a task.  You could make this a part of an extension and use a `ConfigurableFileCollection` on the task. 

As a workaround, I think:
```
srcDir(generatorTask.flatMap { it.outputDirectory })
```
or even
```
srcDir(generatorTask)
```
works.

**Comment by big-guy** [comment]
We don't plan on making any changes to this API to support FileTree specifically.  I described some workarounds/other options in https://github.com/gradle/gradle/issues/27882#issuecomment-1930751306
